### DISCUSSION TITLE
Autocomplete with primitive value and computed label

### DISCUSSION THREAD

**Discussion by zewa666** [ghiscoding/Angular-Slickgrid#1268] [Q&A] [answered]
Hey there @ghiscoding. So these are actually two questions at once.

I'm currently trying out to create a custom backend for Supabase since the query assembly should be hopefully straight forward with the chainable builder approach.

now I have table items with a field called owner_id which is a sequential number which in turn is a foreign key to the id col of the user table.

In order to show the name instead of the owner_id I'd have to make the col an object with key/value pair and set dataKey and labelKey but there is no other way to keep it as number and still have the name rendered right? If not, my approach for that would be a custom formatter which gets the id plus a lookup table passed in and returns the labelKey.

The other question is whether the label can be actually an expression or template to render e.g. "{{firstName}} {{lastName}}. Again in a custom formatter this should be doable but I wanted to check back whether something like this already exists.

**Comment by ghiscoding** [comment, marked as answer]
hello, the Autocomplete lib that I use since removing jQuery/jQueryUI is this lib [kraaden autocomplete](https://github.com/kraaden/autocomplete), it usually works with pair only but there is also the `render` method you can use to change the default display, for that it's better to consult both the 3rd party lib docs and Angular-Slickgrid [Autocomplete Wiki](https://github.com/ghiscoding/angular-slickgrid/wiki/Autocomplete-Editor-(Kraaden-lib)) to see how to use external methods, for example this doc [section](https://github.com/ghiscoding/angular-slickgrid/wiki/Autocomplete-Editor-(Kraaden-lib)#remote-api-renderitem-callback--custom-layout-tworows-or-fourcorners) shows how to modify the render.

> In order to show the name instead of the owner_id I'd have to make the col an object with key/value pair and set dataKey and labelKey but there is no other way to keep it as number and still have the name rendered right? If not, my approach for that would be a custom formatter which gets the id plus a lookup table passed in and returns the labelKey.

yeah the best would be to save it as an object and then use a custom formatter OR you could also use the `Formatters.complex` and then use the `labelKey` and `dataKey` and make sure to set `type: FieldType.object`, there's a demo of that in Example 30, the advantage is that you don't have to create your own custom formatter but you have less flexibility so choose which one is the best approach for you. 

https://github.com/ghiscoding/Angular-Slickgrid/blob/f5ab23ac312922bfa8cd8922bc4dd6f5b375bccf/src/app/examples/grid-composite-editor.component.ts#L306-L323

> The other question is whether the label can be actually an expression or template to render e.g. "{{firstName}} {{lastName}}. Again in a custom formatter this should be doable but I wanted to check back whether something like this already exists.

No we don't have any formatter or any templating engine in the lib, however the beauty of the custom formatters is that you can define your formatter that would handle that, save it as a separate formatter file, and just reuse it across your project. You could use external lib Handlebar or just use RegEx to do it yourself, custom formatter are just plain JS so it should be easy enough :)

**Reply by zewa666** [reply]
alright thanks for the feedback. Yeah eventually I would have gone anyways with a custom formatter and editor in order to have a generic construct for various foreign key relations.

I love this concept really a lot as e.g I'd like to have along with the autocomplete input also a list of "recently used" and the custom editor approach makes this super flexible

also using handlebars in that one to parse the provided template was exactly what I thought 👍
